src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs

```diff
@@ -0,0 +1,187 @@
+
+      FilterDefinition<UnspentOutputTable> unspentOutputFilter =
+         Builders<UnspentOutputTable>.Filter.Eq(utxo => utxo.BlockIndex, blockIndex);
+      Task<DeleteResult> unspentOutput = storage.UnspentOutputTable.DeleteManyAsync(unspentOutputFilter);
+
+      await Task.WhenAll(unspentOutput, output, transactions, addressComputed, addressHistoryComputed, addressUtxoComputed);
+
+      await MergeRewindInputsToUnspentTransactionsAsync(storage,blockIndex);
+
+      FilterDefinition<InputTable> inputFilter =
+         Builders<InputTable>.Filter.Eq(addr => addr.BlockIndex, blockIndex);
+
+      await storage.InputTable.DeleteManyAsync(inputFilter);
+   }
+
+   private static async Task RewindInputDataIntoUnspentTransactionTableAsync(MongoData storage, long blockIndex)
+   {
+      const int limit = 1000;
+      int skip = 0;
+      bool moreItemsToCopy;
+
+      do
+      {
+         var lookupItems = await GetTopNDocumentsFromCollectionAsync<InputTable,InputTable>(storage.InputTable,limit * skip++, limit);
+
+         var itemsToCopy = lookupItems
+            .Where(_ => _.BlockIndex == blockIndex)
+            .ToList();
+
+         var outpoints = itemsToCopy.Select(_ => _.Outpoint).ToList();
+
+         var existingOutpoint = (await storage.UnspentOutputTable
+               .FindAsync(_ => outpoints.Contains(_.Outpoint)))
+            .ToList();
+
+         var filteredItemsToCopy = itemsToCopy
+            .Where(_ => existingOutpoint.All(e => e.Outpoint != _.Outpoint))
+            .Select(_ => new UnspentOutputTable
+            {
+               Address = _.Address, Outpoint = _.Outpoint, Value = _.Value, BlockIndex = -1
+            })
+            .ToList();
+
+         if (!itemsToCopy.Any())
+            break;
+
+         await storage.UnspentOutputTable.InsertManyAsync(filteredItemsToCopy);
+         moreItemsToCopy = itemsToCopy.Count == lookupItems.Count;
+
+
+      } while (moreItemsToCopy );
+   }
+
+   private static async Task DeleteBlockInCollectionFromTopOfTable<T>(IMongoCollection<T> collection,
+      string propertyName, long blockIndex)
+   {
+      const int limit = 10000;
+
+      do
+      {
+         var lookupItems = await GetTopNDocumentsFromCollectionAsync<T,BsonDocument>(collection, 0,limit);
+
+         if (!lookupItems.Any())
+            break;
+
+         foreach (BsonDocument bsonDocument in lookupItems)
+         {
+            if (bsonDocument[propertyName] != blockIndex)
+               return;
+
+            await collection.FindOneAndDeleteAsync(FilterDefinition<T>.Empty,
+               new FindOneAndDeleteOptions<T>
+               {
+                  Sort = new BsonDocumentSortDefinition<T>(new BsonDocument("_id", -1))
+               });
+         }
+
+      } while (true);
+   }
+
+   private static Task MergeRewindInputsToUnspentTransactionsAsync(MongoData storage ,long blockIndex)
```

Comment:
I think perhaps such complex methods need a good description, something like:

"Inputs spend outputs, when an output is spent it gets deleted from the UnspendOutput table and the action of the delete is represented in the inputs table, when a rewind happens we need to bring back outputs that have been deleted from the UnspendOutput so we look for those outputs in the inputs table, however the block index in the inputs table is the one representing the input not the output we are trying to restore so we have to look it up in the outputs table."

Reply:
This got me thinking the inputs we are returning will always be for the current block index no?